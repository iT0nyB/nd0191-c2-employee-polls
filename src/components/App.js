import { useEffect } from "react";
import { connect } from "react-redux";
import { handleInitialData } from "../actions/shared";
import Dashboard from "./Dashboard";
import Container from "react-bootstrap/Container";
import NavigationBar from "./NavigationBar";

const App = (props) => {
  useEffect(() => {
    props.dispatch(handleInitialData());
  }, [props]);
  return (
    <Container className="p-3">
      <NavigationBar />
      <Dashboard />
    </Container>
  );
};

export default connect()(App);
