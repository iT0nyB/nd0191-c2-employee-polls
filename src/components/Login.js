import Button from "react-bootstrap/Button";
import Form from "react-bootstrap/Form";
import Container from "react-bootstrap/Container";
import setAuthedUser from "../actions/authedUser";
import { connect } from "react-redux";
import { useNavigate, useLocation } from "react-router-dom";

import { useState } from "react";

const Login = ({ users, dispatch }) => {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const navigate = useNavigate();

  const location = useLocation();
  const validate = () => {
    return username.length > 0 && password.length > 0;
  };

  const handleLogin = (e) => {
    e.preventDefault();

    if (
      users.find((user) => user.id === username) &&
      users.find((user) => user.password === password)
    ) {
      dispatch(setAuthedUser(username));
      const path = location.pathname;
      if (path.includes("login")) {
        navigate("/");
      } else {
        navigate(path);
      }
    } else {
      alert("Username or Password incorrect");
    }
  };

  return (
    <Container>
      <div className="row">
        <div className="text-center">
          <img
            src="/assets/images/logo192.png"
            alt="Employee Polls Logo"
            width="150"
            className="mx-auto"
          />
          <Form className="w-75 mx-auto">
            <h3>Login to the Employee Polls</h3>
            <Form.Group className="mb-3" controlId="formBasicUsername">
              <Form.Label>Username</Form.Label>
              <Form.Control
                data-testid="username"
                type="username"
                placeholder="Enter username"
                value={username}
                onChange={(e) => setUsername(e.target.value)}
              />
            </Form.Group>

            <Form.Group className="mb-3" controlId="formBasicPassword">
              <Form.Label>Password</Form.Label>
              <Form.Control
                data-testid="password"
                type="password"
                placeholder="Password"
                value={password}
                onChange={(e) => setPassword(e.target.value)}
              />
            </Form.Group>
            <div className="d-grid gap-2">
              <Button
                variant="primary"
                size="lg"
                type="submit"
                onClick={handleLogin}
                data-testid="login-button"
                disabled={!validate()}
              >
                Login
              </Button>
            </div>
          </Form>
        </div>
      </div>
    </Container>
  );
};

const mapStateToProps = ({ users }) => {
  return {
    users: Object.keys(users).map((id) => users[id]),
  };
};

export default connect(mapStateToProps)(Login);
