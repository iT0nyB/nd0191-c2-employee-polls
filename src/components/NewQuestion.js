import Button from "react-bootstrap/Button";
import Form from "react-bootstrap/Form";
import Container from "react-bootstrap/Container";
import { handleAddQuestion } from "../actions/questions";
import { useNavigate } from "react-router-dom";

import { useState } from "react";
import { connect } from "react-redux";

const NewQuestion = (props) => {
  const { authedUser, dispatch } = props;
  const navigate = useNavigate();

  const [optionOneText, setOptionOneText] = useState("");
  const [optionTwoText, setOptionTwoText] = useState("");

  const validate = () => {
    return optionOneText.length > 0 && optionTwoText.length > 0;
  };

  const question = {
    optionOneText,
    optionTwoText,
    author: authedUser,
  };

  const handleNewQuestion = (e) => {
    e.preventDefault();
    dispatch(handleAddQuestion(question));
    setOptionOneText("");
    setOptionTwoText("");
    navigate("/");
  };

  return (
    <Container>
      <div className="row">
        <div className="text-center">
          <Form className="w-75 mx-auto">
            <h3>Would You Rather</h3>
            <p className="muted">Create your own poll</p>
            <Form.Group className="mb-3" controlId="formBasicOptionOne">
              <Form.Label>OptionOne</Form.Label>
              <Form.Control
                type="username"
                placeholder="First Option"
                value={optionOneText}
                onChange={(e) => setOptionOneText(e.target.value)}
              />
            </Form.Group>

            <Form.Group className="mb-3" controlId="formBasicOptionTwo">
              <Form.Label>OptionTwo</Form.Label>
              <Form.Control
                type="optionTwoText"
                placeholder="Second Option"
                value={optionTwoText}
                onChange={(e) => setOptionTwoText(e.target.value)}
              />
            </Form.Group>
            <div>
              <Button
                variant="primary"
                type="submit"
                onClick={handleNewQuestion}
                disabled={!validate}
              >
                Submit
              </Button>
            </div>
          </Form>
        </div>
      </div>
    </Container>
  );
};

const mapStateToProps = ({ authedUser }) => {
  return {
    authedUser: authedUser.id === "" ? null : authedUser,
  };
};

export default connect(mapStateToProps)(NewQuestion);
