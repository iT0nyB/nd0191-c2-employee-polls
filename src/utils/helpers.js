export const formatQuestion = (question, author) => {
  const { id, timestamp, optionOne, optionTwo } = question;
  const { name, avatarURL } = author;

  return {
    name,
    id,
    timestamp,
    avatar: avatarURL,
    optionOne: {
      votes: [],
      text: optionOne,
    },
    optionTwo: {
      votes: [],
      text: optionTwo,
    },
  };
};
