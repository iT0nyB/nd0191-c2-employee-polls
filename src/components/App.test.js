import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import App from "./App";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";

import { legacy_createStore as createStore } from "redux";
import combineReducers from "../reducers";
import applyMiddleware from "../middleware";
import { _saveQuestion, _saveQuestionAnswer } from "../utils/_DATA";
import Login from "../components/Login";
import NewQuestion from "./NewQuestion";
const store = createStore(combineReducers, applyMiddleware);

describe("Login", () => {
  it("will verify that the Login Component is loaded", () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <App />
        </Provider>
      </MemoryRouter>
    );

    expect(screen.getByText("Login to the Employee Polls")).toBeInTheDocument();
  });

  it("will verify that the Login button should be disabled until username and password is entered", () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <App />
        </Provider>
      </MemoryRouter>
    );

    expect(screen.getByText("Login")).toBeDisabled();
  });

  it("will match snapshot", () => {
    const { container } = render(
      <MemoryRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });
});

describe("_saveQuestion", () => {
  it("will verify that the saved question is returned and all expected fields are populated when correctly formatted data is passed to the function.", async () => {
    const mockQues = {
      optionOneText: "Code in VS Code",
      optionTwoText: "Code in Ultra edit",
      author: "abaptiste",
    };
    const savedQuestion = await _saveQuestion(mockQues);
    expect(savedQuestion).toEqual(
      expect.objectContaining({
        id: expect.any(String),
        author: mockQues.author,
        optionOne: { text: mockQues.optionOneText, votes: [] },
        optionTwo: { text: mockQues.optionTwoText, votes: [] },
        timestamp: expect.any(Number),
      })
    );
  });
  it("will verify that an error is returned if incorrect data is passed to the function.", async () => {
    const invalideMockQues = {
      optionTwoText: "Node",
      author: "sarahedo",
    };
    await expect(_saveQuestion(invalideMockQues)).rejects.toEqual(
      "Please provide optionOneText, optionTwoText, and author"
    );
  });
});

describe("_saveQuestionAnswer", () => {
  it("will verify that true is returned when correctly formatted data is passed to the function", async () => {
    const answer = _saveQuestionAnswer({
      authedUser: "sarahedo",
      qid: "vthrdm985a262al8qx3do",
      answer: "optionTwo",
    });

    await expect(answer).resolves.toEqual(true);
  });
  it("will verify that an error is returned if incorrect data is passed to the function", async () => {
    const invalidMockAns = {
      authedUser: "user",
      qid: "qid",
    };
    await expect(_saveQuestionAnswer(invalidMockAns)).rejects.toEqual(
      "Please provide authedUser, qid, and answer"
    );
  });
});

describe("LeaderBoard", () => {
  it("will verify that the Leaderboard is displayed when the link is clicked", () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <App />
        </Provider>
      </MemoryRouter>
    );

    const username = screen.getByTestId("username");
    const password = screen.getByTestId("password");
    expect(screen.getByTestId("login-button")).toBeDisabled();

    fireEvent.change(username, { target: { value: "abaptiste" } });
    fireEvent.change(password, { target: { value: "pass246" } });
    expect(screen.getByTestId("login-button")).not.toBeDisabled();

    const button = screen.getByTestId("login-button");
    fireEvent.click(button);
    expect(screen.getByText("New Questions")).toBeInTheDocument();

    const link = screen.getByTestId("leaderboard-link");
    fireEvent.click(link);
    expect(screen.getByText("LeaderBoard")).toBeInTheDocument();
  });
});

describe("NewQuestion", () => {
  it("will verify that the clicking a new tab will show the new question form", () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <App />
        </Provider>
      </MemoryRouter>
    );

    const link = screen.getByTestId("question-link");

    fireEvent.click(link);
    expect(screen.getByText("Would You Rather")).toBeInTheDocument();
    expect(screen.getByTestId("optionOneText")).toBeInTheDocument();
    expect(screen.getByTestId("optionTwoText")).toBeInTheDocument();
    expect(screen.getByTestId("submit-button")).toBeInTheDocument();
  });
  it("will verify that the submit button on the New Question is disabled until both options are filled", () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <NewQuestion />
        </Provider>
      </MemoryRouter>
    );

    expect(screen.getByTestId("submit-button")).toBeDisabled();

    const option1 = screen.getByTestId("optionOneText");
    const option2 = screen.getByTestId("optionTwoText");

    fireEvent.change(option1, { target: { value: "Test Driven Development" } });
    fireEvent.change(option2, { target: { value: "API First Development" } });

    expect(screen.getByTestId("submit-button")).not.toBeDisabled();
  });
});
