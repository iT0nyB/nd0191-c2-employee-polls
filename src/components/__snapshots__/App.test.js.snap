// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login will match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="text-center"
      >
        <img
          alt="Employee Polls Logo"
          class="mx-auto"
          src="/assets/images/logo192.png"
          width="150"
        />
        <form
          class="w-75 mx-auto"
        >
          <h3>
            Login to the Employee Polls
          </h3>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="formBasicUsername"
            >
              Username
            </label>
            <input
              class="form-control"
              data-testid="username"
              id="formBasicUsername"
              placeholder="Enter username"
              type="username"
              value=""
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="formBasicPassword"
            >
              Password
            </label>
            <input
              class="form-control"
              data-testid="password"
              id="formBasicPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <button
              class="btn btn-primary btn-lg"
              data-testid="login-button"
              disabled=""
              type="submit"
            >
              Login
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
